### PR TITLE
Markdown toolbar: add 'undo' functionality

### PR DESCRIPTION
**Pull request by aitchiss**
## What type of PR is this? (check all applicable)

- [X] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This PR refactors the markdown formatters so that they can now handle undoing a syntax as well as inserting one. This currently only affects the MarkdownToolbar in Storybook, as the toolbar is not yet live in the app.

You'll see when reviewing this one that for the more complex formatters (especially the link one) there's quite a lot of code involved to 'undo' the formatting in a variety of situations. I've tried to lean on the side of "readable chunks", rather than "smallest amount of code", just to be as clear as possible what's being handled at each stage.  

## Related Tickets & Documents

Closes https://github.com/forem/forem/issues/15114

## QA Instructions, Screenshots, Recordings

- Pull down the branch and run `yarn install && yarn storybook`
- Select the `MarkdownToolbar` component from the left navigation to start testing editing features

#### Bold, italic, underline, inline code, strikethrough

- Clicking the related button places the correct formatting on either side of your current selection, keeping your current selection highlighted
- Immediately clicking the button again should remove the formatting on either side, keeping your current selection highlighted
- If you select the whole formatted text - e.g. `**something**` (including all the stars) - and click the button, the formatting on either side should disappear and your selection should now be `something`

#### Links

- Clicking the link button with no text selected, and not inside a current link syntax, should insert the formatting `[](url)` and select the word 'url'
- If you immediately press the link button again (with the word 'url' still highlighted), the whole chunk of formatting should be removed and your cursor should be where you previously inserted the link
- If you select text that begins with either `http://` or `https://` and click the link button, the formatting should appear as `[](http://my-highlighted-url.com)`, and your cursor should now be inside of the square brackets
- If you immediately press the link button again, the formatting should disappear, your link should remain and it should be highlighted/selected
- If you select text that _doesn't_ begin with `http://` or `https://`, the formatting should be inserted as `[my highlighted text](url)` and the url should be highlighted
- If you press the button immediately again, the formatting should be removed and your text description should remain and be highlighted/selected
- If you select a whole chunk of markdown formatting - e.g. `[some text](http://someurl.com)` and click the link button, the formatting should be removed and the description text should remain and be highlighted (i.e. "some text" in this example)
- If you select a whole chunk of markdown formatting which doesn't have a link description - e.g. `[](http://someurl.com)` and click the link button, the formatting should be removed and the link should remain and be highlighted (i.e. "http://someurl.com" in this example)
- If you select a whole chunk of markdown formatting which doesn't have a link description and only has a placeholder url - e.g. `[](url)` and click the link button, the formatting should be removed along with the placeholder link, and your cursor should be where the missing link used to be, with no text selected

#### Headings

- Clicking the heading button once should insert a level 2 heading (ie`## selected text`)
- Clicking it again should change it to a level 3 heading
- Clicking it again should change it to a level 4 heading
- Clicking it again should remove the heading `####`s

#### Ordered list, unordered list, quote

- Clicking the button once should insert the correct markdown on a fresh line
- Clicking it immediately again should remove the line formatting
- If you have multiple lines selected, each line should be formatted or unformatted when you press the button

#### Code block

- Clicking the button should wrap your selected text in the backticks on a new line
- Clicking the button again should remove the backticks on either side of your selection
- If you have no selected text, your cursor is placed inside the empty backticks
- If you select the whole chunk of formatted text (i.e. including the backticks), and press the button again, the backticks should disappear

#### Line divider

- Clicking the button adds a line divider on a fresh line
- Clicking the button again should remove the line divider

### UI accessibility concerns?

Nothing new from the last PR, this is all about how the actually text is manipulated

## Added/updated tests?

- [X] Yes. So many tests 😆 
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [X] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] Are there any post deployment tasks we need to perform?

N/A

## [optional] What gif best describes this PR or how it makes you feel?

Klingon panickedly pushing buttons

### CODE REVIEW

**Comment by jeremyf** [comment]
> I've tried to lean on the side of "readable chunks", rather than "smallest amount of code", just to be as clear as possible what's being handled at each stage.

A great strategy: favor legibility over conciseness.

File: app/javascript/utilities/__tests__/textAreaUtils.test.js
Comment (jeremyf) [inline app/javascript/utilities/__tests__/textAreaUtils.test.js:88]: Really appreciate the legibility of these tests.  I'm rather weak in JS, but this is quite readable for me.

**Comment by jeremyf** [comment]
Regarding headings:

When I start with `## My Heading\n` with my cursor just before the new line, and click the `H` button (Heading).  I get `### My Heading`.  When I click again, I get `#### My Heading`.  When I click again I get `My Heading`.  When I click again I get the following:

```
My Heading

##
```

**Comment by jeremyf** [comment]
Regarding headings:

When I have the line `## My Heading` and my cursor is at the beginning of that line, and I click on the H, then I get the following:

```
##
## My Heading
```

**Comment by jeremyf** [comment]
I have Emacs bindings for my keyboard, I expect <kbd>CTRL</kbd>+<kbd>k</kbd> to delete from the cursor to the end of the line.  Instead I'm getting

My other bindings (e.g. <kbd>CTRL</kbd>+<kbd>d</kbd> for delete character and  <kbd>CTRL</kbd>+<kbd>n</kbd> for next line continue to work).

I don't think this is a show stopper (by any stretch) but wanted to make sure to raise concern.  I don't know best practices regarding application specific keyboard shortcuts.

**Comment by aitchiss** [comment]
> I have Emacs bindings for my keyboard, I expect CTRL+k to delete from the cursor to the end of the line. Instead I'm getting

Oooh interesting! `ctrl/cmd + k` is being used for the 'insert link' shortcut as per the original issue https://github.com/forem/forem/issues/14808. Out of interest - what happens in the github comment text editor when you press ctrl + K? Does it insert a link?

I'm wondering if this is going to cause a bunch of frustration though if people are expecting more default behaviour. CC @ludwiczakpawel @amywtlin

**Comment by jeremyf** [comment]
Regarding lists:  I know that <kbd>*</kbd> is a valid leader for bulleted lists.  In the post that I copied into for testing, I had a list with 2 elements.  Both starting with <kbd>*</kbd>.

When I have:

```

* `article.score`
* `article.hotness_score`

```

And then put my cursor on at the end of the second bullet point, then click to add an unordered list element, I get the following:

```
* `article.score`
* `article.hotness_score`

-
```

With the cursor one space after the newly inserted `-`.

I don't think this is a blocker, but wanted to capture it as it wasn't quite what I was expecting.

**Comment by jeremyf** [comment]
One thing I was partially expecting, but is not part of the stated expected behavior is:

```gherkin
Given I have the following text "this hello thing`
  and my cursor is between the <kbd>e</kbd> and <kbd>l</kbd>
When I click the Bold button
Then the word "hello" is replaced with "**hello**"
```

I don't consider this a blocker but one that could be a future iteration.

**Comment by jeremyf** [comment]
I was partially expecting the Quote function to wrap the text in a Q-tag or even a <kbd>"</kbd>.  What it did was to create a block quote element in markdown.

 I want to consider renaming the tooltip to Blockquote and reconsider the icon.  Both of those prompts lead me to suspect I would be inline quoting something.

**Comment by aitchiss** [comment]
Thanks for all the input so far! I'll wait for more feedback and collate everything together before clarifying/addressing as needed. I think everything so far is expected from my POV (which is not the only POV 😄 ) - the one that probably needs some explanation is headings. 

Currently, when you click the heading button we establish if you are either 'editing a previously inserted heading' (i.e. you have some text selected, but the line starts with `##` already), or you are 'inserting a heading' (i.e. we take a new line and add `##`).

@jeremyf has raised some interesting edge cases that come from this approach that we'll want to think about. Like I say though, please keep the feedback coming at the moment so I can pull together and put some questions to product/design pals if needed 🙂

**Comment by jeremyf** [comment]
> > I have Emacs bindings for my keyboard, I expect CTRL+k to delete from the cursor to the end of the line. Instead I'm getting
> 
> Oooh interesting! `ctrl/cmd + k` is being used for the 'insert link' shortcut as per the original issue #14808. Out of interest - what happens in the github comment text editor when you press ctrl + K? Does it insert a link?
> 
> I'm wondering if this is going to cause a bunch of frustration though if people are expecting more default behaviour. CC @ludwiczakpawel @amywtlin

The <kbd>CTRL</kbd>+<kbd>k</kbd> in Github deletes the line after the cursor.  I'm uncertain what the keyboard shortcuts are for the toolbar of Github.  (For example <kbd>CTRL</kbd>+<kbd>b</kbd> goes backwards one character, as expected in Emacs).

**Comment by aitchiss** [comment]
> The CTRL+k in Github deletes the line after the cursor. I'm uncertain what the keyboard shortcuts are for the toolbar of Github. (For example CTRL+b goes backwards one character, as expected in Emacs).

What I'm gathering here is there's a greater difference between macOS/non-mac functionality than we captured previously. For example `cmd + K` in the github editor inserts a link on macOS. We had incorrectly assumed `ctrl + k` should do the same on other OSs 😬  We'll need to fix this - it'll make the editor experience infuriating for people who use those shortcuts regularly

**Comment by ludwiczakpawel** [comment]
> > I have Emacs bindings for my keyboard, I expect CTRL+k to delete from the cursor to the end of the line. Instead I'm getting
> 
> Oooh interesting! `ctrl/cmd + k` is being used for the 'insert link' shortcut as per the original issue #14808. Out of interest - what happens in the github comment text editor when you press ctrl + K? Does it insert a link?
> 
> I'm wondering if this is going to cause a bunch of frustration though if people are expecting more default behaviour. CC @ludwiczakpawel @amywtlin

I don't think we should be worried about that or even consider this an issue on our side. On github, `CMD/CTRL+K` is used for creating links and only links (if you're in editing mode) - [source](https://docs.github.com/en/get-started/using-github/keyboard-shortcuts). Emacs bindings or any customization happening on user's side is, sorry to say that, user's _problem_. We can't prevent this from happening the same way we can't take responsibility for what browser's 3rd party plugins do with our site 🤷 .

**Comment by aitchiss** [comment]
> I don't think we should be worried about that or even consider this an issue on our side. On github, CMD/CTRL+K is used for creating links and only links (if you're in editing mode) - source. Emacs bindings or any customization happening on user's side is, sorry to say that, user's problem. We can't prevent this from happening the same way we can't take responsibility for what browser's 3rd party plugins do with our site 🤷 .

What I later realised is that for mac users, we're overriding the `ctrl` key commands as well as adding the `cmd` ones (not great for users like @jeremyf who use the readline keybindings 😅 ). I've added a separate PR for that because it's kind of embedded in our `KeyboardShortcuts` component: https://github.com/forem/forem/pull/15265

**Comment by citizen428** [comment]
Amazing work @aitchiss 👏 

Some observations:

* When we accidentally select spaces before or after the Markdown doesn't get removed but applied a second time (e.g. accidentally select `**text** ` will result in `****text** **`). I know this is technically a user error but I wonder if we could potentially ignore leading and trailing whitespace?
* When I select a word and make it a heading two empty newlines get added before it:
    * Selection: image
    * After pressing `H`: image
    * When I cycle from 4 `#` to none and then click `H` again, more empty lines will be added. image
* Inserting ordered or unordered lists also adds empty lines, both with and without text selected

**Comment by citizen428** [comment]
> Emacs bindings [...] is, sorry to say that, user's problem.

@ludwiczakpawel Not really. Every macOS text input responds to Emacs key bindings (`Ctrl-A` beginning of line, `Ctrl-E` end of line, `Ctrl-K` forward delete to end of line) by default and has done so for a long time, at least all the way back to when I got my first Intel Mac in 2007. See [Mac keyboard shortcuts])(https://support.apple.com/en-us/HT201236), section "Document shortcuts".

**Comment by amywtlin** [comment]
In terms of Emacs key binding, I need to research a bit more as I am not familiar in what circumstances someone would consider using it (maybe @jeremyf can shed some light on this?). I am also looking to get a sense of the adoption, to gauge how much it would impact our immediate users (DEV users essentially) based on the nature of the audience (i.e. do a lot of developers use Emacs key binding? If so we probably need to address it).

Here's the general thinking process (ok, I am basically rephrasing the above, in an ordered list):
1. What personas would use Emacs key bindings?
2. How does the adoption of Emacs key binding look like? (I quickly googled `emacs adoption` but nothing much shows up...) I am looking for percentages across major browsers.


Please point me to any relevant sources you know of that would help answer these questions!

**Comment by citizen428** [comment]
> In terms of Emacs key binding, I need to research a bit more as I am not familiar in what circumstances someone would consider using it

Like I said, they work in every macOS text input and amongst other things provide functionality that Mac keyboards don't have dedicated keys for (like "Home" and "End" in a full keyboard).

**Comment by amywtlin** [comment]
I agree that it is ideal to respect the original setting in one's OS. To clarify, I think this is a good problem to solve. I wonder though in what urgency level should we deal with this. 

If a lot of developers use Emac keys binding, then it's possible it becomes a blocker of the feature (which I suspect would not be the case, based on the fact we only realize there could potentially be issues with Emac binding users). If there are some but not a lot, then we can ship first and think about how we can address this.

**Comment by amywtlin** [comment]
I created a [poll](https://forem-team.slack.com/archives/C7GE84DEJ/p1635920303229000) to use our own developers as a proxy of the wider DEV audience to learn more about the behaviors 🤩 Let's see what people have to say

**Comment by citizen428** [comment]
Anyway, the only one that seems to really be problematic is `Ctrl-k` which should forward delete, not insert a URL. I concede that probably few enough macOS users are aware of this shortcut AND using it regularly to not be a massive problem. 

I feel maybe a bit too strongly about this because I had/have problems with RSI, so I'm very attached to keyboard heavy workflows which is e.g. why I use Vimium instead of a mouse in my browser.

**Comment by amywtlin** [comment]
@citizen428 I think this PR by Suzanne should solve the main issue here:
>What I later realised is that for mac users, we're overriding the ctrl key commands as well as adding the cmd ones (not great for users like @jeremyf who use the readline keybindings 😅 ). I've added a separate PR for that because it's kind of embedded in our KeyboardShortcuts component: #15265

The culprit here is we override `ctrl-k` also for macOS users, where we should only treat `ctrl-k` as adding a link for Windows users. If I understand correctly, after this PR is merged, macOS users can happily use `ctrl-k` for forward delete again, while `cmd-k` remains the action of adding a link 👯 (@aitchiss did I understand this correctly though?)

**Comment by ludwiczakpawel** [comment]
> Every macOS text input (including e.g. browser address bars) implements Emacs key bindings [1] by default and has done so for a long time, at least all the way back to when I got my first Intel Mac in 2007. See Mac keyboard shortcuts, section "Document shortcuts".

@citizen428 what I actually meant (but likely failed at expressing) was that we should use as native and standard behaviors as possible.

For macOS: use **only** `CMD+K` for links, and leave CTRL+K untouched (so it can keep/inherit its default behavior)
For Windows: use only `CTRL+K` for links.

In other words, if we've been overwriting some default keyboard shortcuts by _accident_ (i.e. using `CMD+K` **and** `CTRL+K` on macOS for inserting links) then that's bad, but looks like #15265 will fix it 👍 .

Unless there are some other misunderstandings here I think we should close _this_ topic - it's fairly simple (to explain) issue that escalated pretty quickly causing unnecessary discussion (I blame myself for that as well 😞 ).

**Comment by aitchiss** [comment]
Thanks everyone for the feedback so far ❤️ 

I'm going to summarise where we're at, so I can get advice on what's a blocker, what's not, and highlight what's being addressed elsewhere.

@amywtlin @ludwiczakpawel - would be great if I can get a product steer on the open questions about what goes in to this version and what's noted for 'future enhancements'.

## Keyboard shortcuts

**(Status: solved ✅ )**

First up, apologies from me, because my initial misreading of the comment on this I think stirred up more confusion 😂 

The issue at hand is that on a mac, both `ctrl + <letters>` executes the same formatter as `cmd + <letters>`. This bug is baked into an existing component, and therefore I've resolved it separately in https://github.com/forem/forem/pull/15265. Once this is merged, we should no longer have this problem.

## Heading level formatter

**(Status: questions ❓)**

- When text isn't selected (i.e. I just drop my cursor on a line), there are two different behaviours depending on whether the line already has `#`s
- This means that if you have no text selected and cycle through all the heading levels, after level 4 it will stop incrementing the current line, and assume you want to insert a new heading
- The video below probably explains a lot better than I can in words 😄 

https://user-images.githubusercontent.com/20773163/140033468-89436b18-597b-4b6f-9698-cf01b84ef385.mp4

- This only happens if you don't have any text selected

My question is: is this a blocker? What should it do? I'm wary of treating every empty text selection as "edit this whole line", because I don't think it's what people want. For example, at the beginning of the video, the heading is inserted on a new line from "Line two of text" and I think that is what users would want.


## New (empty) lines being added

**(Status: questions ❓)**

As flagged up by @citizen428, when we insert block level syntax, we add 2 new lines before the block syntax, and one after. This was by design based on previous feedback from @ludwiczakpawel.

We discussed at the time that the 'new line' logic could be "smarter" - detecting if there are already 2 new lines preceding the current text selection or not, and only inserting the new lines if not. We decided previously to keep things simple, but based on Michael's feedback, do we want to reconsider?

I don't think the logic would be too challenging to add but it is another layer of complexity.

## Unordered list formatter

**(Status: questions ❓)**

@jeremyf rightly highlighted that the unordered list formatter inserts and edits lists created with `-` only, and not `*`. Naturally, from an 'insert' point of view this is absolutely fine, but if a user manually types a list with `*` bullets, the formatter will not work to "undo" them.

The question is - are we happy with this limitation?

For what it's worth, I lean strongly in favour of accepting this limitation. I don't think there is a need to support multiple ways of adding/removing the same syntax, and I can see random edge cases popping up if we try to handle both.

## Formatters only format the selected text

**(Status: questions ❓)**

@jeremyf highlighted that if I have my cursor in the middle of a word, and click an inline formatter like `bold`, the word itself is not formatted bold - i.e. this doesn't happen:

```
Given I have the following text "this hello thing`
And my cursor is between the 'e' and 'l'
When I click the Bold button
Then the word "hello" is replaced with "**hello**"
```

We already have a similar point on some automatic behaviours depending on cursor position added to the 'future improvements' list over here: https://github.com/forem/forem/issues/14808#issuecomment-946488867. Are we happy to leave this for a future improvement too?

## Quote vs Blockquote

**(Status: questions ❓)**

Feedback:

>I was partially expecting the Quote function to wrap the text in a Q-tag or even a ". What it did was to create a block quote element in markdown.

>I want to consider renaming the tooltip to Blockquote and reconsider the icon. Both of those prompts lead me to suspect I would be inline quoting something.

Happy to rename this - as for icon, @ludwiczakpawel I defer to you 🙂 


## Trailing whitespaces

**(Status: questions ❓)**

> When we accidentally select spaces before or after the Markdown doesn't get removed but applied a second time (e.g. accidentally selecting `**text** ` will result in `****text** **`). I know this is technically a user error but I wonder if we could potentially ignore leading and trailing whitespace?

I'm naturally biased on a lot of this, so I'm happy for product to advise, but I don't think we should solve for this. My feeling is that the "undo" functionality is primarily to solve for the case where:

- User inserts syntax
- User realises they made a mistake and immediately presses again

What we have at the moment is working for a lot more use cases than e.g. the github toolbar/shortcuts, and I'm wary to stray to far and add in too much magic 😄 

But like I say, I am very biased here, so would love a pointer from product/the team on this one.

**Comment by citizen428** [comment]
> As flagged up by @citizen428, when we insert block level syntax, we add 2 new lines before the block syntax, and one after. This was by design based on previous feedback from @ludwiczakpawel.

I think this just was a bit irritating because it happened while being on the first line and I guess this would commonly be a heading. So you start typing into a completely empty text box, select the text, then hit `H` and suddenly everything moves down 2 lines, it just *feels* like a bug, even if it was intended.

> What we have at the moment is working for a lot more use cases than e.g. the github toolbar/shortcuts

Absolutely. I just noticed it and figured I'll leave the feedback so that it's at least known but I fully agree that we don't need to handle edge cases like accidentally selected whitespace.

**Comment by jeremyf** [comment]
> ## Keyboard shortcuts
> **(Status: solved ✅ )**

Huzzah!

> ## Heading level formatter
> **(Status: questions ❓)**
> 
> My question is: is this a blocker? What should it do? I'm wary of treating every empty text selection as "edit this whole line", because I don't think it's what people want. For example, at the beginning of the video, the heading is inserted on a new line from "Line two of text" and I think that is what users would want.

I don't consider this a blocker.  I think I now see this as "Let's release this for people to test and see how they respond".

> ## New (empty) lines being added
> **(Status: questions ❓)**
> 
> We discussed at the time that the 'new line' logic could be "smarter" - detecting if there are already 2 new lines preceding the current text selection or not, and only inserting the new lines if not. We decided previously to keep things simple, but based on Michael's feedback, do we want to reconsider?
> 
> I don't think the logic would be too challenging to add but it is another layer of complexity.

I think this is at your discretion.  You could create an issue to capture this additional work.  With that I'd say we've covered our bases.

> 
> ## Unordered list formatter
> **(Status: questions ❓)**
> 
> @jeremyf rightly highlighted that the unordered list formatter inserts and edits lists created with `-` only, and not `*`. Naturally, from an 'insert' point of view this is absolutely fine, but if a user manually types a list with `*` bullets, the formatter will not work to "undo" them.
> 
> The question is - are we happy with this limitation?
> 
> For what it's worth, I lean strongly in favour of accepting this limitation. I don't think there is a need to support multiple ways of adding/removing the same syntax, and I can see random edge cases popping up if we try to handle both.

I trust your experienced intuition on edge cases popping up.  So, let's keep this limitation, and we don't even need an issue (unless we get user feedback saying it's an issue)

> 
> ## Formatters only format the selected text
> **(Status: questions ❓)**
> 
> @jeremyf highlighted that if I have my cursor in the middle of a word, and click an inline formatter like `bold`, the word itself is not formatted bold - i.e. this doesn't happen:
> 
> ```
> Given I have the following text "this hello thing`
> And my cursor is between the 'e' and 'l'
> When I click the Bold button
> Then the word "hello" is replaced with "**hello**"
> ```
> 
> We already have a similar point on some automatic behaviours depending on cursor position added to the 'future improvements' list over here: [#14808 (comment)](https://github.com/forem/forem/issues/14808#issuecomment-946488867). Are we happy to leave this for a future improvement too?

Again, happy with this being a future improvement.

> ## Quote vs Blockquote
> **(Status: questions ❓)**
> Happy to rename this - as for icon, @ludwiczakpawel I defer to you 🙂

I'll away @ludwiczakpawel's response, and defer to their experience and recommendation.

> ## Trailing whitespaces
> **(Status: questions ❓)**
> 
> > When we accidentally select spaces before or after the Markdown doesn't get removed but applied a second time (e.g. accidentally selecting `**text** ` will result in `****text** **`). I know this is technically a user error but I wonder if we could potentially ignore leading and trailing whitespace?
> 
> I'm naturally biased on a lot of this, so I'm happy for product to advise, but I don't think we should solve for this. My feeling is that the "undo" functionality is primarily to solve for the case where:
> 
> * User inserts syntax
> * User realises they made a mistake and immediately presses again
> 
> What we have at the moment is working for a lot more use cases than e.g. the github toolbar/shortcuts, and I'm wary to stray to far and add in too much magic 😄
> 
> But like I say, I am very biased here, so would love a pointer from product/the team on this one.

Again, sounds like this is not a blocker for this PR, but would be a natural starting point for an iterative improvement.

**Comment by amywtlin** [comment]
@aitchiss Thanks for the summary! 

I have to say I agree with @jeremyf's comments on each of them here. I don't think any of them is a blocker, and I think your intuition is aligned with my thinking. 

Re: block/blockquote, I think for `quote` I would expect this:
> This is is a quote.

So the current behavior is what I would have expected... I actually think the name and the icon is fine. Plus Q-tag is a HTML tag, if I understand correctly. I feel `>` insertion is the correct behavior upon hitting `quote` button. Looks like GitHub also calls this a "quote."

If we don't feel strongly about it, I'd say ship it as is and let's see if we get complaints.

**Comment by aitchiss** [comment]
Thank you everyone for the input on those questions!

I've updated my previous comment to note the new statuses, and raised issues etc where needed. None of the feedback so far has been identified as blocking, so this PR is still very much looking for reviews 🙂

**Comment by jeremyf** [comment]
> So the current behavior is what I would have expected... I actually think the name and the icon is fine. Plus Q-tag is a HTML tag, if I understand correctly. I feel `>` insertion is the correct behavior upon hitting `quote` button. Looks like GitHub also calls this a "quote."

For context, Q-tag is an inline tag (e.g., when you use it there's no line breaks) whereas BLOCKQUOTE-tag is a block tag (e.g., there are line breaks).  The icon of Github implies indentation.  However, none of this is a blocker, mostly my goal to clarify in regards to the interplay of semantic HTML and UI/UX.

**Comment by jeremyf** [comment]
@aitchiss for the issues that you created from these conversations, could you make sure to reference this PR.  It may be helpful for the next portion of the work to be able to go back and refer to these conversations.

**Review by jeremyf — APPROVED** [review]
This is a tremendous step forward on the Toolbar.  Great work!

**Review by nickytonline — COMMENTED** [review]
This is working great!

Somewhat related as I was curious about the undo for a list.

For some options, like an unordered list, are we considering shortcuts? I tried CMD + Shift + 7 like in https://docs.github.com/en/get-started/using-github/keyboard-shortcuts#source-code-editing, but it didn't work.

**Comment by aitchiss** [comment]
> For some options, like an unordered list, are we considering shortcuts?

Good question @nickytonline and the answer is - yup! But we don't have a shortcut for the lists I'm afraid. I should have posted this in the original PR description, but the shortcuts are:

- Bold (cmd/ctrl+B)
- Italic (cmd/ctrl+I)
- Link (cmd/ctrl+K)
- Underline (cmd/ctrl+U)
- Strikethrough (cmd/ctrl+shift+X)

When you hover a toolbar button (or focus it) you should see the shortcut command if it exists 🙂

**Comment by nickytonline** [comment]
> > For some options, like an unordered list, are we considering shortcuts?
> 
> 
> 
> Good question @nickytonline and the answer is - yup! But we don't have a shortcut for the lists I'm afraid. I should have posted this in the original PR description, but the shortcuts are:
> 
> 
> 
> - Bold (cmd/ctrl+B)
> 
> - Italic (cmd/ctrl+I)
> 
> - Link (cmd/ctrl+K)
> 
> - Underline (cmd/ctrl+U)
> 
> - Strikethrough (cmd/ctrl+shift+X)
> 
> 
> 
> When you hover a toolbar button (or focus it) you should see the shortcut command if it exists 🙂 

Yeah, I saw the shortcuts on hover, was just curious about the lists, etc. Thanks!

**Review by nickytonline — APPROVED** [review]
Great stuff @aitchiss!

**Pull request merged**
